### ISSUE TITLE
All tests fail with Regex=[FAILED]

### ISSUE THREAD

**Issue by afonso360** [mozilla/rr#2017]
I'm trying to build mozilla rr on Arch linux, if this is not a supported OS for some reason please close this issue.

I had a similiar error (same output) but reading LastTest.log revealed that ptrace was failing so I executed
`echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope` and the error messages changed.

I suspected this might be an issue of arch using python3 by default, so I tried creating a python2.7 virtual environment and running the tests with that active, however it still produced the same error.

I'm using ninja to build rr and when I execute `ninja test` all tests fail with the following error message (Only tested the first 30 or so)
```
[0/1] Running tests...
Test project /home/afonso/git/rr
        Start   1: 64bit_child
  1/2048 Test   #1: 64bit_child .................................................***Failed  Error regular expression found in output. Regex=[FAILED]  0.82 sec
        Start   2: 64bit_child-no-syscallbuf
  2/2048 Test   #2: 64bit_child-no-syscallbuf ...................................***Failed  Error regular expression found in output. Regex=[FAILED]  1.54 sec
        Start   3: _llseek
```

Here's the output of [`perf list`](https://pastebin.mozilla.org/9018438)

I had to shorten this log to 1000 lines because pastebin refused more: [`LastTest.log.tmp`](https://pastebin.mozilla.org/9018440)

```
~/git/rr>$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 69
model name      : Intel(R) Core(TM) i7-4510U CPU @ 2.00GHz
stepping        : 1
microcode       : 0x1f
cpu MHz         : 1122.741
cache size      : 4096 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perf$
on pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 movbe popcnt tsc_deadlin$
_timer aes xsave avx f16c rdrand lahf_lm abm epb tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt dtherm ida arat pln pts
bugs            :
bogomips        : 5188.14
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:

<<Sniped>>
```

```
~/git/rr>$ dmesg|grep PMU
[    0.153941] Performance Events: PEBS fmt2+, Haswell events, 16-deep LBR, full-width counters, Intel PMU driver.
[    0.154542] NMI watchdog: enabled on all CPUs, permanently consumes one hw-PMU counter.
[    0.684285] core: PMU erratum BJ122, BV98, HSD29 worked around, HT is on
[    7.405708] RAPL PMU: API unit is 2^-32 Joules, 4 fixed counters, 655360 ms ovfl timer
[    7.405710] RAPL PMU: hw unit of domain pp0-core 2^-14 Joules
[    7.405711] RAPL PMU: hw unit of domain package 2^-14 Joules
[    7.405711] RAPL PMU: hw unit of domain dram 2^-14 Joules
[    7.405712] RAPL PMU: hw unit of domain pp1-gpu 2^-14 Joules
```

**Comment by khuey** [comment]
The stack printed from one of the tests is

```
Thread 1 (Thread 0x7f83f8583b80 (LWP 2270)):
#0  0x00007f83f6ff6410 in __nanosleep_nocancel () from /usr/lib/libc.so.6
#1  0x00007f83f6ff636a in sleep () from /usr/lib/libc.so.6
#2  0x000000000074870c in rr::notifying_abort () at /home/afonso/git/rr/src/util.cc:960
#3  0x0000000000631a52 in rr::FatalOstream::~FatalOstream (this=0x7ffdb9086230) at /home/afonso/git/rr/src/log.cc:270
#4  0x0000000000656b3b in rr::RecordSession::create (argv=std::vector of length 1, capacity 8 = {...}, extra_env=std::vector of length 0, capacity 0,
    syscallbuf=rr::RecordSession::ENABLE_SYSCALL_BUF, bind_cpu=-2) at /home/afonso/git/rr/src/RecordSession.cc:1733
#5  0x000000000064ebf9 in rr::record (args=std::vector of length 1, capacity 8 = {...}, flags=...) at /home/afonso/git/rr/src/RecordCommand.cc:333
#6  0x000000000064dd17 in rr::RecordCommand::run (this=0xa1d518 <rr::RecordCommand::singleton>, args=std::vector of length 1, capacity 8 = {...})
    at /home/afonso/git/rr/src/RecordCommand.cc:462
#7  0x000000000074b708 in main (argc=7, argv=0x7ffdb90866e8) at /home/afonso/git/rr/src/main.cc:270
```

And the relevant frame (4) has

```
    ScopedFd fd("/proc/sys/kernel/perf_event_paranoid", O_RDONLY);
    if (fd.is_open()) {
      char buf[100];
      ssize_t size = read(fd, buf, sizeof(buf) - 1);
      if (size >= 0) {
        buf[size] = 0;
        int val = atoi(buf);
        if (val > 1) {
          FATAL() << "rr needs /proc/sys/kernel/perf_event_paranoid <= 1, but "
                     "it is "
                  << val << ".\nChange it to 1, or use 'rr record -n' (slow).\n"
                  << "Consider putting 'kernel.perf_event_paranoid = 1' in "
                     "/etc/sysctl.conf";
        }
    }
```

Are we not printing that message if you run rr on something trivial outside the test harness (e.g. /bin/true)?

**Comment by afonso360** [comment]
Yes it does print that message.
I then used `echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid` and that does
make the message disappear, but the tests are still failing with the same error.

I also tested with `perf_event_paranoid = 0` but the result is still the same

Here's the new [`LastTest.log.tmp`](https://pastebin.mozilla.org/9018441)

The output of running rr on /bin/true is this:
```
~/git/rr>$ ./bin/rr record /bin/true

rr: Warning: Your CPU frequency governor is 'powersave'.  rr strongly
    recommends that you use the 'performance' governor.  Not using the
    'performance' governor can cause rr to be at least 2x slower
    on laptops.

    On Fedora-based systems, you can enable the 'performance' governor
    by running the following commands:

    $ sudo dnf install kernel-tools
    $ sudo cpupower frequency-set -g performance

rr: Saving execution to trace directory `/home/afonso/.local/share/rr/true-3'.
```

I also tried executing  `sudo cpupower frequency-set -g performance` and rerunning the tests, it still fails.

**Comment by khuey** [comment]
Can you replay that recording (`rr replay -a`)?

**Comment by afonso360** [comment]
Not really

```
~/git/rr>$ ./bin/rr replay -a
[FATAL /home/afonso/git/rr/src/replay_syscall.cc:473:process_execve() errno: ENOENT]
 (task 7623 (rec:7125) at time 14)
 -> Assertion `false' failed to hold. Exec of stub /home/afonso/git/rr/bin/../bin/rr_exec_stub
failed
Launch gdb with
  gdb '-l' '10000' '-ex' 'target extended-remote :7623'
[FATAL /home/afonso/git/rr/src/log.cc:296:emergency_debug() errno: SUCCESS] Can't resume execut
ion from invalid state
Aborted (core dumped)
```

gdb output: 
```
~>$ gdb '-l' '10000' '-ex' 'target extended-remote :7623'
GNU gdb (GDB) 7.12.1
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Remote debugging using :7623
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x00000000005862b6 in ?? ()
(gdb) break main
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (main) pending.
(gdb) c
Continuing.
Remote connection closed
(gdb)
```
Heres the output of coredumpctl when i extracted the core dump file
```
 ~>$ coredumpctl dump -o ~/git/rr/rr.core.dump
           PID: 7622 (rr)
           UID: 1000 (afonso)
           GID: 100 (users)
        Signal: 6 (ABRT)
     Timestamp: Sun 2017-04-09 12:39:48 UTC (4min 58s ago)
  Command Line: ./bin/rr replay -a
    Executable: /home/afonso/git/rr/bin/rr
 Control Group: /user.slice/user-1000.slice/session-c2.scope
          Unit: session-c2.scope
         Slice: user-1000.slice
       Session: c2
     Owner UID: 1000 (afonso)
       Boot ID: cfb02ec0bdbb4232bc19766cfe16d4a8
    Machine ID: 756b837f924840149ad0935fa1aed82c
      Hostname: AYYLMAO
       Storage: /var/lib/systemd/coredump/core.rr.1000.cfb02ec0bdbb4232bc19766cfe16d4a8.7622.1491741588000000000000.lz4
       Message: Process 7622 (rr) of user 1000 dumped core.

                Stack trace of thread 7622:
                #0  0x00007ff9dc669a10 raise (libc.so.6)
                #1  0x00007ff9dc66b13a abort (libc.so.6)
                #2  0x0000000000748714 _ZN2rr15notifying_abortEv (rr)
                #3  0x0000000000631a52 _ZN2rr12FatalOstreamD1Ev (rr)
                #4  0x0000000000631f90 _ZN2rrL15emergency_debugEPNS_4TaskE (rr)
                #5  0x0000000000631da2 _ZN2rr21EmergencyDebugOstreamD1Ev (rr)
                #6  0x00000000006d6621 _ZN2rrL14process_execveEPNS_10ReplayTaskERKNS_10TraceFrameEPNS_15ReplayTraceStepE (rr)
                #7  0x00000000006d50a8 _ZN2rrL24rep_process_syscall_archINS_7X64ArchEEEvPNS_10ReplayTaskEPNS_15ReplayTraceStepERKNS_9RegistersE (rr)
                #8  0x00000000006d35b6 _ZZN2rr19rep_process_syscallEPNS_10ReplayTaskEPNS_15ReplayTraceStepEENK3$_0clERKNS_9RegistersE (rr)
                #9  0x00000000006d071e _ZN2rr24with_converted_registersIvZNS_19rep_process_syscallEPNS_10ReplayTaskEPNS_15ReplayTraceStepEE3$_0EET_RKNS_9RegistersENS_13SupportedArchET0_ (rr)
                #10 0x00000000006d0612 _ZN2rr19rep_process_syscallEPNS_10ReplayTaskEPNS_15ReplayTraceStepE (rr)
                #11 0x00000000006cb1c8 _ZN2rr13ReplaySession28setup_replay_one_trace_frameEPNS_10ReplayTaskE (rr)
                #12 0x00000000006cb4d7 _ZN2rr13ReplaySession11replay_stepERKNS0_15StepConstraintsE (rr)
                #13 0x00000000006c38b5 _ZN2rr13ReplaySession11replay_stepENS_10RunCommandE (rr)
                #14 0x00000000006c2c88 _ZN2rrL24serve_replay_no_debuggerERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_11ReplayFlagsE (rr)
                #15 0x00000000006c1f05 _ZN2rrL6replayERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_11ReplayFlagsE (rr)
                #16 0x00000000006c11df _ZN2rr13ReplayCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE (rr)
                #17 0x000000000074b708 main (rr)
                #18 0x00007ff9dc656511 __libc_start_main (libc.so.6)
                #19 0x000000000057d4da _start (rr)
More than one entry matches, ignoring rest.
```
If you need the file, tell me on how i can send it to you

**Comment by khuey** [comment]
Ok.  Can you figure out why the exec stub isn't there?

**Comment by afonso360** [comment]
Running the executable `./bin/rr_exec_stub` caused the following error to show up
```
~/git/rr/>$ ./bin/rr_exec_stub 
bash: ./bin/rr_exec_stub: No such file or directory
```

Running `readelf -a ./bin/rr_exec_stub` yielded the following line
`[Requesting program interpreter: /lib/ld64.so.1]`
which does not appear to be in my system, so I symlinked it
`sudo ln -s /lib64/ld-linux-x86-64.so.2 /lib/ld64.so.1`

I did a similar thing for the 32bit version
`sudo ln -s /lib/ld-linux.so.2 /usr/lib/libc.so.1`

I searched in my package manager for packages including those two files but couldn't find any (wierd), ill need to investigate further

I can now execute `./bin/rr_exec_stub` and `./bin/rr_exec_stub_32` however it imediatley segfaults
coredumpctl doesen't show much information

```
~/git/rr>$ ./bin/rr_exec_stub
Segmentation fault (core dumped)
~/git/rr>$ coredumpctl dump
           PID: 6547 (rr_exec_stub)
           UID: 1000 (afonso)
           GID: 100 (users)
        Signal: 11 (SEGV)
     Timestamp: Sun 2017-04-09 15:25:25 UTC (45s ago)
  Command Line: ./bin/rr_exec_stub
    Executable: /home/afonso/git/rr/bin/rr_exec_stub
 Control Group: /user.slice/user-1000.slice/session-c2.scope
          Unit: session-c2.scope
         Slice: user-1000.slice
       Session: c2
     Owner UID: 1000 (afonso)
       Boot ID: cfb02ec0bdbb4232bc19766cfe16d4a8
    Machine ID: 756b837f924840149ad0935fa1aed82c
      Hostname: AYYLMAO
       Storage: /var/lib/systemd/coredump/core.rr_exec_stub.1000.cfb02ec0bdbb4232bc19766cfe16d4a8.6547.1491751525000000000000.lz4
       Message: Process 6547 (rr_exec_stub) of user 1000 dumped core.

                Stack trace of thread 6547:
                #0  0x0000000000000001 n/a (n/a)
Refusing to dump core to tty (use shell redirection or specify --output).
```
But It now executes successfully all tests and `rr replay -a` works

**Comment by rocallahan** [comment]
> I can now execute ./bin/rr_exec_stub and ./bin/rr_exec_stub_32 however it imediatley segfaults

That's expected.

> But It now executes successfully all tests and rr replay -a works

Great.

**Comment by rocallahan** [comment]
I'm not sure what the underlying problem is here. We build `rr_exec_stub` with `-static -nostartfiles -nodefaultlibs` (and `-fno-stack-protector`) so it should not be dynamically linked and should not mention any form of `ld.so`.

**Comment by afonso360** [comment]
Would you perfer me to leave this issue open and we can discuss that, or should I open a new issue?

**Comment by rocallahan** [comment]
I'll reopen it.

**Comment by rocallahan** [comment]
Try
````
touch ./src/exec_stub.c
VERBOSE=1 make
````
And show us what you get

**Comment by afonso360** [comment]
Ok so this is a bit wierd, I downloaded a fresh copy of rr, compiled with make as per the instructions, then did `touch` and `make` with verbose, this is the [log file](https://pastebin.mozilla.org/9018473) for the exec_stub section, which shows the correct build flags, and running `readelf -a` does not show the dependency on `/lib/ld64.so`

I then recompiled the whole thing in a difrent directory (including cmake generation) and built that with ninja, heres the [`ninja -v` log](https://pastebin.mozilla.org/9018515) which also shows the correct flags, but the resulting executable has the `ld64.so` dependency.

[make `readelf -a ./bin/rr_exec_stub`](https://pastebin.mozilla.org/9018516)
[ninja `readelf -a ./bin/rr_exec_stub`](https://pastebin.mozilla.org/9018519)

**Comment by rocallahan** [comment]
If you delete `rr_exec_stub` and then rebuild it by running the `clang` command from the command line, do you get the nonstatic executable?

If so you should be able to reduce the command by removing unnecessary command line parameters (most of them aren't relevant) to narrow down what's going wrong.

Since this is a link step you might want to figure out what linker is being used and try running it directly.

**Comment by afonso360** [comment]
Yes, rebuilding with the clang command directly yields the same result.

Removing -rdynamic causes the produced binary to become static.

My knowledge here is limited you might have to help me a bit more with this

**Comment by rocallahan** [comment]
We shouldn't be passing `-rdynamic`, we don't need it. I guess that's confusing your version of clang somehow. I'll see if I can remove it.
